### ISSUE TITLE
Unexpected performance degradation on GSM8K when multi-turn/tool use is enabled compared to disabled

### ISSUE THREAD

**Issue by eshoyuan** [volcengine/verl#1569]
Firstly, congratulations on the exciting release of the new multi-turn RLHF framework with tool support! The capabilities it brings are a significant step forward for Agentic RL training.

I've been experimenting with the framework based on your recent post on Zhihu 
SGLang verl OpenBMB 与清华大学团队联合开源：在主流 RLHF 框架上首次支持多轮交互与工具调用 - Chayenne Zhao的文章 - 知乎
https://zhuanlan.zhihu.com/p/1906007821889283171

I successfully ran the Qwen2.5-3B-Instruct model on the GSM8K task using the provided `run_qwen2.5-3b_gsm8k_multiturn.sh` script. My results on the validation set were a `mean@1` score of approximately 0.82, which is consistent with the `wandb` log you shared.

To understand the impact of the new features, I then conducted an experiment where I disabled multi-turn and tool use. Specifically, I modified the configuration as follows:

```yaml
actor_rollout_ref:
  rollout:
    name: "sglang_async"
    multi_turn:
      enable: false
      tool_config_path: null # Explicitly set to null
````

All other settings, including the train/validation data split for GSM8K, remained identical. With multi-turn and tool use disabled, the `mean@1` score on the GSM8K validation set improved to 0.86.

This result was quite surprising to me. The Gsm8kTool provided in the example is, in essence, a very powerful oracle, as it evaluates the correctness of the math problem's answer based on the ground truth to calculate the reward. Theoretically, leveraging such a strong tool should lead to significantly better performance compared to not using it, as it provides very accurate feedback. However, my experiment shows the opposite.

Config Diff: https://www.diffchecker.com/lJr4Xw2x/

**Comment by lebronjamesking** [comment]
Yes, I ran into same puzzled situation. I think in the run_qwen2.5-3b_gsm8k_multiturn.sh script provided, the model_max_length=1024 which is way less than expected 'multi-turn' of tool calling. I guess something 10240 or so is needed in order to improve the gsm8k competency. BTW, **qwen2.5-0.5b-instruct** along achive over 83% on mean1. So normally above 83% is expected during grpo rl.

**Comment by eshoyuan** [comment]
> Yes, I ran into same puzzled situation. I think in the **run_qwen2.5-3b_gsm8k_multiturn.sh** script provided, the model_max_length=1024 which is way less than expected 'multi-turn' of tool calling. In other words, you don't get any chance of tool calling by limiting such short response length. The model is still struggling with step by step reasoning before actually calling any tool. I guess something 10240 or so is needed in order to improve the gsm8k competency. BTW, **qwen2.5-0.5b-instruct** along achive over 83% on mean1. So normally above 83% is expected during grpo rl.
> 
> Anyway, hopefully you need multiple nodes with each 8_A100 or 8_H20 cards to achieve improved gsm8k result.

I completely agree that a model_max_length of 1024 is quite short for multi-turn tool calling.

I tried increasing the response length to 4096, and while the mean1 score went up to 84%, it was still lower than without the tool. The average response length hovered around 600 before the training crashed,  similar to the issue reported [here](https://github.com/volcengine/verl/issues/1581).

Even though 1024 feels restrictive, theoretically, the model should be able to learn to improve accuracy within those limitations if everything is working correctly.

<img width="1279" alt="Image" src="https://github.com/user-attachments/assets/1cdd5632-6947-4111-8897-e889b6847e67" />

<img width="410" alt="Image" src="https://github.com/user-attachments/assets/24b9d9d2-84dd-4a87-9c6c-b4363ed40faf" />

**Comment by Kirrito-k423** [comment]
中文：我不建议现在直接关闭这个 issue。现有公开证据更像是“原始多轮工具 quickstart 的训练信号/轨迹行为需要进一步确认”，而不是一个已经被明确验证修复的单点 bug。

我看到一个相关的已合并 PR：#4998（https://github.com/verl-project/verl/pull/4998）。该 PR 明确引用了本 issue，并给出一个复现实验解释：官方 multi-turn + Gsm8kTool 配置中，`turns` 指标会很快退化到 2，模型基本跳过 `<tool_call>` 直接猜最终答案；在加入“包含 `<tool_call>` 时 +0.1”的 reward shaping 后，实验中的验证准确率从约 0.608 提升到约 0.717。这支持一个可能原因：工具本身能返回强反馈，但当前训练目标未必会稳定诱导模型实际调用工具。

另外，原示例里 `data.max_response_length=1024` 对多轮推理/工具调用确实偏紧；相关讨论里也提到增加到 4096 后仍未完全超过 no-tool baseline，并出现训练稳定性问题。相关 issue #1581 中也有维护者建议切换到 AgentLoopManager/token-in-token-out 路径，以避免旧 chat/message 往返造成 token 不一致。

如果继续追这个问题，建议补充或复测这些指标：使用的 verl commit、完整脚本/config diff、是否仍使用旧 `sglang_async`/ChatScheduler 路径还是 AgentLoopManager、`num_turns`、tool-call count/rate、`response_length/clip_ratio`、是否发生 NaN/grad_norm spike，以及有无 tool-call shaping/custom reward。

English: I would not close this issue yet. The public evidence points more to “the original multi-turn tool quickstart needs more trajectory/training-signal validation” than to a confirmed, fully fixed single bug.

I found a related merged PR, #4998 (https://github.com/verl-project/verl/pull/4998), which explicitly references this issue. Its reproduction says the official multi-turn + Gsm8kTool setup quickly collapses to `turns = 2`, meaning the model mostly skips `<tool_call>` and guesses the final answer directly. With a small reward-shaping bonus for trajectories containing `<tool_call>`, the reported validation accuracy improved from about 0.608 to about 0.717. This supports a likely explanation: the tool can provide strong feedback, but the training objective may not reliably incentivize the model to call it.

Also, the original example’s `data.max_response_length=1024` is likely tight for multi-turn reasoning/tool use. The follow-up discussion says increasing it to 4096 improved the score but still did not fully beat the no-tool baseline and was associated with training instability. In related issue #1581, maintainers also recommended moving to the AgentLoopManager/token-in-token-out path to avoid token inconsistency from older chat/message round trips.

For next debugging, it would help to share or rerun with: the exact verl commit, full script/config diff, whether the run uses legacy `sglang_async`/ChatScheduler or AgentLoopManager, `num_turns`, tool-call count/rate, `response_length/clip_ratio`, any NaN/grad_norm spike, and whether tool-call shaping/custom reward is enabled.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
